Check that plugin name is kebab-cased upon generation

WalkingPizza opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the documentation. (Should be made against the documentation branch)
- Create or update the tests.
- Refer to the issue you are closing in the PR description - fix #issue
- Specify if the PR is in WIP (work in progress) state or ready to be merged

Please ensure you read through the Contributing Guide: https://github.com/strapi/strapi/blob/master/CONTRIBUTING.md
-->

### What does it do?

Checks that the plugin the user is about to generate has a kebab-cased name.

### Why is it needed?

Strapi only allows kebab-cased plugin names, but the `generate` utility does not check that this requirement is met upon plugin generation.

### How to test it?

Try generating a plugin with a non-kebab-cased name.

### Related issue(s)/PR(s)

Solves #12311

markkaylor reviewed (commented):
@WalkingPizza this suggestion looks good.

I'm just wondering if we could format the user input ourself so that it is cased the way we want in the template file.

Maybe we could do that with inquirer somehow by transforming the input or use [case modifiers](https://plopjs.com/documentation/#case-modifiers) in the template file directly.

I think that would be interesting and we could also add `displayName` property to the package json strapi object using `titleCase`

Let me know what you think @WalkingPizza, if you agree with this approach and are willing to make the change that would be great. Otherwise we can merge this and I can explore that myself later on.

WalkingPizza commented:
Hey @markkaylor!

I think transforming the user input automatically is a very good idea, but what would we do when a user tries to generate a plugin with a non-kebab-cased name which converts to something that is already taken?

E.g.
1. User generates the `sample-plugin` plugin.
2. User generates the `samplePlugin` plugin, whose name is then automatically converted to `sample-plugin`, which is already taken.

I 100% agree with the automatic `displayName` property too, and I am more than happy to make the changes before this is merged.

markkaylor commented:
As it is right now if I generate a plugin with the same name the CLI throws an error. The generator should generate file paths in kebab case as well and then I think it wouldn't be possible to generate a plugin with the same name regardless of user input, right? Let me know if you spot something I'm not seeing. 

- User generates `sample-plugin` => `path/to/sample-plugin`
- User generates. `samplePlugin` (converted to sample-plugin) => `path/to/sample-plugin`
- CLI throws: `File already exists, abort` 

Here's what I'm thinking:

```hbs
// templates/plugin-package.json.hbs
 "name": "{{ pluginName }}", // kebab-case
  "version": "0.0.0",
  "description": "This is the description of the plugin.",
  "strapi": {
    "name": "{{ pluginName }}", // kebab-case
    "description": "Description of {{ pluginName }} plugin", // Title Case
    "kind": "plugin"
    "displayName":: {{ pluginName }} // Title Case
  },
```

```javascript
// plops/plugin.js
{
  type: 'add',
  path: 'plugins/{{ pluginName }}/package.json', // kebab-case
  templateFile: 'templates/plugin-package.json.hbs',
},
```

WalkingPizza commented:
@markkaylor I have:

1. Added a utility in `@strapi/utils` to convert strings to kebab-case since Lodash is not accessible directly inside the generators.
2. Added a warning when users try to generate a plugin whose name is automatically converted to kebab-case to notify them that Strapi only supports kebab-cased names and that the new plugin has been automatically renamed.
3. Added the displayName property to the newly generated `package.json`.

Let me know what you think!

markkaylor reviewed (changes requested):
Awesome, this works great! Thanks @WalkingPizza, I just have one small comment and then I think we can approve and merge.

WalkingPizza commented on packages/generators/generators/lib/templates/plugin-package.json.hbs:
```diff
@@ -5,7 +5,8 @@
   "strapi": {
     "name": "{{ pluginName }}",
     "description": "Description of {{ pluginName }} plugin",
```
Done!

derrickmehaffy commented:
@WalkingPizza can you resolve the conflicts?

alexandrebodin reviewed (approved):
Thank @WalkingPizza 💯
